frame/contracts/proc-macro/src/lib.rs

```diff
@@ -140,3 +147,407 @@ fn format_default(field: &Ident) -> TokenStream {
 		&self.#field
 	}
 }
+
+/// Parsed environment definition.
+struct EnvDef {
+	host_funcs: Vec<HostFn>,
+}
+
+/// Parsed host function definition.
+struct HostFn {
+	item: syn::ItemFn,
+	module: String,
+	name: String,
+	returns: HostFnReturn,
+}
+
+enum HostFnReturn {
+	Unit,
+	U32,
+	ReturnCode,
+}
+
+/// Helper trait to convert a host function definition into its wasm signature.
+trait ToWasmSig {
+	fn to_wasm_sig(&self) -> TokenStream;
+}
+
+impl ToWasmSig for HostFn {
+	fn to_wasm_sig(&self) -> TokenStream {
+		let args = self.item.sig.inputs.iter().skip(1).filter_map(|a| match a {
+			syn::FnArg::Typed(pt) => Some(&pt.ty),
+			_ => None,
+		});
+		let returns = match &self.returns {
+			HostFnReturn::U32 => quote! { vec![ <u32>::VALUE_TYPE ] },
+			HostFnReturn::ReturnCode => quote! { vec![ <ReturnCode>::VALUE_TYPE ] },
+			HostFnReturn::Unit => quote! { vec![] },
+		};
+
+		quote! {
+			 wasm_instrument::parity_wasm::elements::FunctionType::new(
+				vec! [ #(<#args>::VALUE_TYPE),* ],
+				#returns,
+			)
+		}
+	}
+}
+
+impl ToTokens for HostFn {
+	fn to_tokens(&self, tokens: &mut TokenStream) {
+		self.item.to_tokens(tokens);
+	}
+}
+
+impl HostFn {
+	pub fn try_from(item: syn::Item) -> syn::Result<Self> {
+		let err = |span, msg| {
+			let msg = format!("Invalid host function definition. {}", msg);
+			syn::Error::new(span, msg)
+		};
+		let msg = "only #[version(<u8>)] or #[unstable] attribute is allowed.";
+		let span = item.span();
+		let item = match item {
+			syn::Item::Fn(i_fn) => Ok(i_fn),
+			_ => Err(err(span, msg)),
+		}?;
+
+		let name = item.sig.ident.to_string();
+		let mut module = "seal0".to_string();
+		let attrs: Vec<&syn::Attribute> =
+			item.attrs.iter().filter(|m| !m.path.is_ident("doc")).collect();
+
+		match attrs.len() {
+			0 => (),
+			1 => {
+				let attr = &attrs[0];
+				let ident = attr.path.get_ident().ok_or(err(span, msg))?.to_string();
+				match ident.as_str() {
+					"version" => {
+						let ver: syn::LitInt = attr.parse_args()?;
+						module = format!(
+							"seal{}",
+							ver.base10_parse::<u8>().map_err(|_| err(span, msg))?
+						);
+					},
+					"unstable" => {
+						module = "__unstable__".to_string();
+					},
+					_ => return Err(err(span, msg)),
+				}
+			},
+			_ => return Err(err(span, msg)),
+		}
```

Comment:
Possibly a matter of personal style but I find the following more understandable: 

```suggestion
        let module = 
		match attrs.len() {
			0 => Ok("seal0".to_string()),
			1 => {
				let attr = &attrs[0];
				let ident = attr.path.get_ident().ok_or(err(span, msg))?.to_string();
				match ident.as_str() {
					"version" => {
						let ver: syn::LitInt = attr.parse_args()?;
						Ok(format!(
							"seal{}",
							ver.base10_parse::<u8>().map_err(|_| err(span, msg))?
						))
					},
					"unstable" => {
						Ok("__unstable__".to_string())
					},
					_ => Err(err(span, msg)),
				}
			},
			_ => Err(err(span, msg)),
		}?;
```